### PR TITLE
change: ETCM-7811 native token management pallet

### PR DESCRIPTION
**Pull request by AmbientTea**
# Description

The pallet that consumes inherent data about number of tokens moved into illiquid supply on main chain and calls a Partner Chain's custom handler.

This is the last PR of the native token observability. You can see the changes from all the PR together in [this draft PR](https://github.com/input-output-hk/partner-chains/pull/22). 

# Checklist

- [x] Commit sequence broadly makes sense and commits have useful messages.
- [x] New tests are added if needed and existing tests are updated.
- [x] Relevant logging and metrics added
- [x] CI passes. See note on CI.
- [x] Any changes are noted in the `changelog.md` for affected crate
- [x] Self-reviewed the diff

# Note on CI
If your PR is from a fork, the necessary CI jobs won't trigger automatically for security reasons.
You will need to get someone with write privileges.  Please contact IOG Partner Chains developers to do this
for you.

### CODE REVIEW

File: Cargo.toml [hunk 1]
```diff
@@ -34,6 +34,7 @@ members = [
     "primitives/session-manager",
     "primitives/sidechain",
 	"partner-chains-cli",
+    "pallets/native-token-management",
```
Comment (LGLO) [inline Cargo.toml:37]: How do we "manage" this token? Is it the business/product name of the feature?
I know that it is quite late, for this comment. Maybe change could be a separate PR now.

File: devnet/.envrc [hunk 2]
```diff
@@ -15,6 +15,11 @@ export COMMITTEE_CANDIDATE_ADDRESS=$(jq -r '.addresses.CommitteeCandidateValidat
 export D_PARAMETER_POLICY_ID=$(jq -r '.mintingPolicies.DParameterPolicy' devnet/addresses.json)
 export PERMISSIONED_CANDIDATES_POLICY_ID=$(jq -r '.mintingPolicies.PermissionedCandidatesPolicy' devnet/addresses.json)
 
+# native token observability
+export PARTNER_CHAIN_NATIVE_TOKEN_POLICY_ID='ada83ddd029614381f00e28de0922ab0dec6983ea9dd29ae20eef9b4'
+export PARTNER_CHAIN_NATIVE_TOKEN_ASSET_NAME='5043546f6b656e44656d6f'
+export PARTNER_CHAIN_ILLIQUID_SUPPLY_VALIDATOR_ADDRESS='addr_test1wrhvtvx3f0g9wv9rx8kfqc60jva3e07nqujk2cspekv4mqs9rjdvz'
```
Comment (LGLO) [inline devnet/.envrc:21]: This will be inconsistent. We don't use PARTNER_CHAIN_ prefix for our configs. There is exception to BLOCK_BENEFICIARY, but as you can see it didn't aged well.

File: node/src/template_chain_spec.rs [hunk 3]
```diff
@@ -44,6 +44,18 @@ pub fn chain_spec() -> Result<ChainSpec, EnvVarReadError> {
 			initial_authorities: vec![],
 			main_chain_scripts: read_mainchain_scripts_from_env()?,
 		},
+		native_token_management: NativeTokenManagementConfig {
+			main_chain_scripts: sp_native_token_management::MainChainScripts {
```
Comment (LGLO) [inline node/src/template_chain_spec.rs:48]: Is it possible to have `read_main_chains_scripts` in `sp_native_token_management`, that doesn't depend on `sc_*`, so it would be used here like `main_chain_scripts: sp_native_token_management::read_main_chains_scripts` ?

This would adhere to the approach when we move as much code to library crates as possible.

File: pallets/native-token-management/src/lib.rs [hunk 4]
```diff
@@ -0,0 +1,175 @@
...
+				.filter(|data| !data.token_amount.0 > 0)
+				.map(|data| Call::transfer_tokens { token_amount: data.token_amount })
+		}
+
+		fn check_inherent(call: &Self::Call, data: &InherentData) -> Result<(), Self::Error> {
+			let Some(token_transfer) = Self::get_transfered_tokens_from_inherent_data(data) else {
+				return Err(InherentError::UnexpectedTokenTransferInherent);
+			};
+			let Call::transfer_tokens { token_amount } = call else {
+				unreachable!("There is no other extrinsic in this pallet");
```
Comment (LGLO) [inline pallets/native-token-management/src/lib.rs:108]: `set_main_chain_scripts` is the other extrinsic in this pallet.

File: pallets/native-token-management/src/lib.rs [hunk 5]
```diff
@@ -0,0 +1,175 @@
...
+				.map(|_| InherentError::TokenTransferNotHandled))
+		}
+	}
+
+	impl<T: Config> Pallet<T> {
+		fn get_transfered_tokens_from_inherent_data(
+			data: &InherentData,
+		) -> Option<TokenTransferData> {
+			data.get_data::<TokenTransferData>(&INHERENT_IDENTIFIER)
+				.expect("Token transfer data not correctly encoded")
```
Comment (LGLO) [inline pallets/native-token-management/src/lib.rs:132]:
```suggestion
				.expect("Token transfer data is not encoded correctly")
```

File: pallets/native-token-management/src/mock.rs [hunk 6]
```diff
@@ -0,0 +1,92 @@
+use crate::mock::sp_runtime::testing::H256;
+use crate::DispatchResult;
+use crate::TokenTransferHandler;
+use frame_support::sp_runtime::traits::BlakeTwo256;
+use frame_support::sp_runtime::traits::IdentityLookup;
+use frame_support::sp_runtime::BuildStorage;
+use frame_support::traits::ConstU16;
+use frame_support::traits::ConstU32;
+use frame_support::traits::ConstU64;
```
Comment (LGLO) [inline pallets/native-token-management/src/mock.rs:9]: Cosmetic: I don't like wildcard imports, to be honest, but okay. However, "paid by LOC" style of imports is even worse.

File: Cargo.toml (hunk 1 above)
Comment (AmbientTea) [reply Cargo.toml:37]: The requirements doc calls the feature "native token management". I don't think it's a good name either. Since there's already crates with the management name merged, let's eventually revisit it in a separate PR as you said.

File: pallets/native-token-management/src/lib.rs (hunk 4 above)
Comment (AmbientTea) [reply pallets/native-token-management/src/lib.rs]: Right, I added it later. Using `match` would have made the compiler catch that change. changed that to return `Ok(())` like other `check_inherent`s

File: pallets/native-token-management/src/lib.rs (hunk 5 above)
Comment (AmbientTea) [reply pallets/native-token-management/src/lib.rs]: Applied

File: pallets/native-token-management/src/lib.rs [hunk 7]
```diff
@@ -0,0 +1,175 @@
...
+
+	#[pallet::inherent]
+	impl<T: Config> ProvideInherent for Pallet<T> {
+		type Call = Call<T>;
+		type Error = InherentError;
+		const INHERENT_IDENTIFIER: InherentIdentifier = INHERENT_IDENTIFIER;
+
+		fn create_inherent(data: &InherentData) -> Option<Self::Call> {
+			Self::get_transfered_tokens_from_inherent_data(data)
+				.filter(|data| !data.token_amount.0 > 0)
```
[the remark this answers is not in the document]
Comment (AmbientTea) [reply pallets/native-token-management/src/lib.rs]: 🤦 right

File: pallets/native-token-management/src/lib.rs [hunk 8]
```diff
@@ -0,0 +1,175 @@
...
+use sp_native_token_management::*;
+
+#[cfg(test)]
+mod tests;
+
+#[cfg(any(test, feature = "mock"))]
+mod mock;
+
+#[derive(Encode, Debug, PartialEq)]
+pub enum InherentError {
```
[the remark this answers is not in the document]
Comment (AmbientTea) [reply pallets/native-token-management/src/lib.rs:35]: You're absolutely right, I didn't realize that error in IDP is the inherent's error not IDP's error

[the remark this answers is not in the document]
Comment (AmbientTea) [reply pallets/native-token-management/src/lib.rs]: Fixed that

File: devnet/.envrc (hunk 2 above)
Comment (AmbientTea) [reply devnet/.envrc]: You're right, I removed it

File: node/src/template_chain_spec.rs (hunk 3 above)
Comment (AmbientTea) [reply node/src/template_chain_spec.rs]: I added a helper function in the node crate for now. We have one for the other main chain scripts and the env var reading helpers are here as well, so I din't want to move too much stuff around.

**Review by mpskowron — APPROVED** [review]

File: node/src/template_chain_spec.rs (hunk 3 above)
Comment (LGLO) [reply node/src/template_chain_spec.rs]: I'm strongly for having as much code as possible out of node. I don't say to move other. Leave others for a specific PR with refactor, but lets not add to node if it's relatively easy, please.

If we merge it here and make a release, users will have to C-P it. If we move it out then, users will have to follow refactor.

**Review by LGLO — APPROVED** [review]
